### ISSUE TITLE
`slice::get_mut()` followed by `slice::copy_from_slice` generates unreachable panic branch

### ISSUE THREAD

**Issue by bugadani** [rust-lang/rust#98294]
In code like the following, the compiler misses a possible optimization: in `f1`, the length of `dst` is equal to the length of `bytes`, yet the compiler generates a call to `len_mismatch_fail`.

```rust
pub fn f1(a: &mut [u8], offset: usize, bytes: &[u8]) {
    if let Some(dst) = a.get_mut(offset..offset + bytes.len()) {
        dst.copy_from_slice(bytes);
    }
}
```

The compiler knows the lengths are equal (and thus, the panic is unreachable), because the following snippet optimizes the panic away:

```rust
pub fn f2(a: &mut [u8], offset: usize, bytes: &[u8]) {
    if let Some(dst) = a.get_mut(offset..offset + bytes.len()) {
        assert!(dst.len() == bytes.len());
        dst.copy_from_slice(bytes);
    }
}
```

This is a regression between rustc versions 1.51 and 1.52.

https://rust.godbolt.org/z/YhEY78E9P

**Comment by steffahn** [comment]
This seems to be fixed on nightly. @rustbot label E-needs-test

Bisection would still be interesting, both for the regression and the fix.

**Comment by steffahn** [comment]
Alright, that’s pretty boring in terms of rustc-bisection: it was fixed by #99464 (Update to LLVM 15), and it regressed in `nightly-2021-03-05` which includes #81451 (Upgrade[^1] to LLVM 12).

@rustbot label -E-needs-bisection, +A-llvm.

[^1]: what’s up with the inconsistent naming, upgrade vs. update?
